roadmap/implementers-guide/src/runtime/disputes2.md

```diff
@@ -0,0 +1,54 @@
+# Disputes
+
+## Motivation
+
+If the transaction can be verified the defined verification code, it is good. If not, there is a dispute
+about a block, which made it through the backing stage, but was then during secondary checking found
+to be invalid.
+At this point the validator set can not be trusted (since they voted for the block despite something being
+fishy at the very least) or the validator that blew the whistle has ulterior motives to do so.
+In either way, there are malicious validators around.
+As a consequence, all validators at the time of block backing, are being notified via broadcast of the first challenging vote.
+Validators that backed the candidate implicitly voted for backing the candidate and may not vote a second time.
+That would be equal to an attempted double vote and would be slashed accordingly.
+
+## Initiation
+
+A dispute is initiated by one secondary checker creating and gossiping a vote, that challenges the vote.
+Secondary checkers run the validation code and use this to verify the transactions.
+
+After a secondary checker challenged a block, all validators that received the gossiped vote, check the
+block themselves via the validation code.
+The result of that check is converted into a vote, and distributed via the same mechanics as the first one.
+
+Once a receiver receives quorum and supermajority in one or the other direction, the
+vote is concluded.
+Conclusion implies that the result for this block can not be altered anymore, valid or invalid is fixed now.
+
+In order to assure, the dispute result is not forgotten. It has to be recorded on chain.
+This on chain recording mechanic must be vigilant, in a sense, that new emerging forks
+must also receive the dispute resolution recorded (transplantation) iff the disputed block is container there or is in the
+ancestry path.
+
+If the disputed block was already finalized, the block must be put in governance mode to be resolved be by hand
+(i.e. sudo or motion or other mechanics that are available ).
+
+Now there might be the case that the network is fragmented for a limited amount of time. If the dispute would happen
+during this time, a block could be disputed, that we have never seen being part of any chain, but we might get
+a vote for it. That is called a "remote" dispute.
+
+As such the validator has to keep track of all votes irrespective if the disputed block is already known or not.
+All backing votes should be either kept in storage as well, or be queried on demand, since they are a kind of vote
+as well.
```

Comment:
I'd say the dispute should contain enough data to prove that all validators should begin processing the dispute, meaning at least one backing vote and at least one disputing vote.  

We're resolving using the validator set at inclusion time, so the backing vote could actually come from another validator set, but that's okay because the backers' stake remains locked and they remain liable.

Reply:
What would happen if noone is backing the proposed block?

Reply:
Good point!  If you receive an invalid candidate from a collator then what should you do?   It's not a dispute because no validator voted for the candidate yet.  It's almost surely buggy parachain code or buggy collator deployment, but maybe someone has launched an attack and just sent the bad candidate to you by mistake?  

There are two solutions, either (a) we hang onto the candidate until another candidate from that parachain gets finalized, or else (b) we throw it away and figure we can reconstruct ourselves if required, or else (c) ignore the problem and leave it to approval checkers.

I suspect Rob might dislike (a) so maybe we could implement (b) saying "backers have the right to reconstruct a block like tranche zero approval checkers".  In this way, you simply ignore invalid blocks that nobody else backs when backing, but then recheck under this bizarre situation.  We then later partially do (a) by saying that we deprioritize storing the block but do not actually flush out its storage yet, but this adds nothing formally. 

In the short term, I'd vote (c) ignore unbacked invalid blocks unless they're backed, but later maybe we'll try (b) or even this partial (a).

Reply:
I think we've got to accept disputes on anything that at least one validator has issued a backing statement for. Otherwise we'd have to mess around with trying to prove to some remote chain whether a block was actually backed by a majority of its backing validators, and who those validators were, etc.

Reply:
Well, now that I think about it, proving this to a remote chain might not actually be that hard assuming that every chain has the same view of sessions / validators / paras. But that is kind of a stretch as we don't fix paras one session in advance the same way we do validators/randomness. It is a change I am open to, though.

Reply:
But the implications of noone backing that particular block would be no inclusion, as such why would we care to deal with it any further down the row, instead of just dropping the negative vote? I think we need to talk some more about the details of that.

Reply:
Actually my (b) sounds pointless now, so our approach goes roughly:

We've a backer who obtains a candidate B but discovers B to be invalid.  Initially, our backer does nothing but ideally^2 retains B for a little while.  If our backer ever sees a backing vote for B then our backer should ideally^1 create a dispute, and retain B permanently if they have B.  Ideally^3 our disputing backer should then participate in erasure coded piece distribution.  Ideally^4 if a backer appears absent then we ask for one more approval checker, but do not count the missing backer as a no-show.

We've three simplifications obtained by removing bits of functionality prefixed by ideally^i for i>k.  All four ideally^i add only small-ish improvements so I do not want to push for them when people have higher priority things.  If an ideally^i looks really easy to do, then go for it.  :)  If we add none of these ideally^i then backers simply ignore invalid candidates.

Reply:
Ideally^1 & Ideally^2 seem reasonable to impl, not sure about ^3 and ^4